### PR TITLE
IDVA6-544 incorporate api calls into the frontend

### PR DESCRIPTION
**Pull request by mmodzelewski-ch**
Link to Jira:
https://companieshouse.atlassian.net/browse/IDVA6-544

Changes:
- private-api-sdk-node package added to the project;
- associations service created
- code refactored to use associations service instead of mocked data
- code refactored for various improvements

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/routers/controllers/confirmCompanyController.ts [hunk 1]
```diff
@@ -17,10 +17,9 @@ export const confirmCompanyControllerGet = async (req: Request, res: Response):
 export const confirmCompanyControllerPost = async (req: Request, res: Response): Promise<void> => {
     const session: Session = req.session as Session;
     const company = session.data.extra_data.companyProfile;
-    const userEmailAddress = getLoggedInUserEmail(req.session);
     const companyNotActive = company.companyStatus.toLocaleLowerCase() !== constants.COMPANY_STATUS_ACTIVE;
 
-    const isAlreadyAssociated = await isCompanyAssociatedWithUser(company.companyNumber, userEmailAddress);
+    const isAlreadyAssociated = await isCompanyAssociatedWithUser(req, company.companyNumber);
```
Comment (kkonuganti-ch) [inline src/routers/controllers/confirmCompanyController.ts:22]: is already associated can be done using the new endpoint which takes email as filter.

/associations/{company_number}?yser_email=<email>

File: src/routers/controllers/resendEmailController.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { Request, Response } from "express";
 import * as constants from "../../constants";
-import { isEmailAuthorised } from "../../services/userCompanyAssociationService";
+import { isEmailAuthorised } from "../../services/associationsService";
 import { getExtraData, setExtraData } from "../../lib/utils/sessionUtils";
 import { sendAuthorisationEmail } from "../../services/emailNotificationService";
```
Comment (kkonuganti-ch) [inline src/routers/controllers/resendEmailController.ts:5]: what does this do?
i though emails will be sent by backend.

File: src/routers/controllers/resendEmailController.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ export const resendEmailController = async (req: Request, res: Response): Promis
     const validEmail = validateEmailString(email);
     let emailOnList = false;
     if (validEmail) {
-        emailOnList = await isEmailAuthorised(email, companyNumber);
+        emailOnList = await isEmailAuthorised(req, email, companyNumber);
```
Comment (kkonuganti-ch) [inline src/routers/controllers/resendEmailController.ts:15]: how are you checking this, if email is in the list as the data returned by end point is paginated. If more then 15 we would not be able to check the data. hence i have created a new filter on /associations/compamy_number endpoint/

File: src/routers/controllers/confirmCompanyController.ts (hunk 1 above)
Comment (mmodzelewski-ch) [reply src/routers/controllers/confirmCompanyController.ts:22]: Do you mean /associations/companies/{company_number}?user_email=

File: src/services/associationsService.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        logger.error(`Associations API returned no resource for associations with status ${JSON.stringify(status)}`);
+        return Promise.reject(sdkResponse);
+    }
+
+    logger.debug(`Received associations ${JSON.stringify(sdkResponse)}`);
+
+    return Promise.resolve(sdkResponse.resource as Associations);
+};
+
+export const isCompanyAssociatedWithUser = async (req: Request, companyNumber: string): Promise<string> => {
```
Comment (kkonuganti-ch) [inline src/services/associationsService.ts:35]: is this a get /associations or get /associations/{company_number}. if latter then you can pass user email as parameer.

File: src/routers/controllers/confirmCompanyController.ts (hunk 1 above)
Comment (mmodzelewski-ch) [reply src/routers/controllers/confirmCompanyController.ts:22]: I've changed the isCompanyAssociatedWithUser function so it uses /associations/companies/{company_number}?user_email= endpoint

File: src/routers/controllers/resendEmailController.ts (hunk 2 above)
Comment (mmodzelewski-ch) [reply src/routers/controllers/resendEmailController.ts:5]: At the moment it does nothing, just returns hard coded status 201. Although the job of resending email notification may be done by the backend once it's implemented, the frontend has to initiate the whole process when the appropriate link is clicked and then display confirmation that the email has been resent based on the response from the backend. So at the moment this is just a skeleton function to be filled with logic once this part of the system is ready.

File: src/routers/controllers/resendEmailController.ts (hunk 3 above)
Comment (mmodzelewski-ch) [reply src/routers/controllers/resendEmailController.ts:15]: Yes, this function uses /associations/companies/{company_number}?user_email= endpoint

File: src/services/associationsService.ts (hunk 4 above)
Comment (mmodzelewski-ch) [reply src/services/associationsService.ts:35]: I modified this function to use /associations/companies/{company_number}?user_email= endpoint

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/services/associationsService.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+export const getUserAssociations = async (req: Request, status: AssociationStatus[]): Promise<Associations> => {
+    const apiClient = createOauthPrivateApiClient(req);
+
+    logger.info(`Looking for associations with status ${JSON.stringify(status)}`);
+    const sdkResponse: Resource<Associations | Errors> = await apiClient.associationsService.searchAssociations(status);
+
+    if (!sdkResponse) {
+        logger.error(`Associations API for status ${JSON.stringify(status)}`);
+        return Promise.reject(sdkResponse);
+    }
```
Comment (bmchugh-ch) [inline src/services/associationsService.ts:18]: Probably not something we need to handle now but we will need to update when we are working on the unhappy path / error handling.

Normally promise.rejection takes a rejection reason why the promise is rejected or an error. 
In this instance we are just passing undefined as the rejection reason.

**Review by bmchugh-ch — APPROVED** [review]
Good work

File: src/routers/controllers/resendEmailController.ts (hunk 2 above)
Comment (kkonuganti-ch) [reply src/routers/controllers/resendEmailController.ts:5]: this should call the invitation endpoint and back end will take care of things.

**Review by kkonuganti-ch — APPROVED** [review]
LGTM

File: src/routers/controllers/resendEmailController.ts (hunk 2 above)
Comment (mmodzelewski-ch) [reply src/routers/controllers/resendEmailController.ts:5]: At the time this code was created, there was nothing to call and we had a very vague understanding of how this feature is supposed to work. I assume there is an integration ticket now that will allow us to add the missing logic.

**Pull request merged**